mondey_backend/src/mondey_backend/import_data/manager/import_manager.py

```diff
@@ -0,0 +1,952 @@
+                        parent_id = self.get_childs_parent_id(
+                            session, child_row.get("CASE")
+                        )
+                        _, answer = self.create_answer(
+                            session,
+                            user_or_child_id=parent_id,
+                            question_id=question.id,
+                            answer_text=answer_text,
+                            set_only_additional_answer=set_only_additional_answer,
+                            is_child_question=False,
+                        )
+                    else:
+                        # Create child answer
+                        _, answer = self.create_answer(
+                            session,
+                            user_or_child_id=db_child_id,
+                            question_id=question.id,
+                            answer_text=answer_text,
+                            set_only_additional_answer=set_only_additional_answer,
+                            is_child_question=True,
+                        )
+
+                    total_answers += 1
+                    session.add(answer)
+
+                else:
+                    logger.warning(
+                        f"Variable type has no clear processing method: {variable_type}, {variable_label}"
+                    )
+                    missing += 1
+
+        session.commit()
+        logger.info(f"Total answers saved: {total_answers}")
+        logger.info(f"Missing answers: {missing}")
+
+    async def import_additional_data(self, session: Session) -> None:
+        """Import additional data.
+        @param session is so we can overwrite this for tests (like test_import_additional.py)"""
+        logger.info("Importing additional data")
+
+        if not self.data_manager.import_paths.additional_data_path:
+            logger.warning("No additional data path provided")
+            return
+
+        additional_data_df = self.data_manager.load_additional_data_df()
+
+        # First import children with milestone data
+        await self.import_children_with_milestone_data(session, additional_data_df)
+
+        # Then import answers
+        self.import_answers(session, additional_data_df)
+
+        logger.info("Additional data imported successfully")
+
+    # -------------------------------------------------------------------------
+    # Import Method
+    # -------------------------------------------------------------------------
+
+    async def run_additional_data_import(self) -> None:
+        """Run import of additional data with DB rollback for any errors."""
+        logger.info("Starting additional data import")
+
+        # Get sessions
+        import_session, _ = self.data_manager.get_import_session()
+
+        try:
+            # Import additional data
+            await self.import_additional_data(import_session)
+
+            logger.info("Additional data import completed successfully")
+
```

Comment:
My understanding is that rollback only rolls back uncommited changes - once you commit there is no going back.

So I think we would need to not do *any* commits in the code here, then do a single commit after import_additional_data ran succcessfully, and otherwise catch any exception and rollback (as done currently)

https://docs.sqlalchemy.org/en/20/orm/session_basics.html#framing-out-a-begin-commit-rollback-block

Reply:
Maybe replacing all the `.commit` calls with `.flush` would work?

Reply:
Also a related point: I would pass in the session and user_session as arguments everywhere and then call this code from the router passing in the session and user_session. If you instead create a session within this code, then once it goes out of scope `.commit` is implicitly called, so I think it's safer to avoid doing this if we want to be able to rollback all the changes. (Also we should rollback the user_session as well)